modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -17,94 +17,128 @@
+
 import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import java.util.concurrent.ExecutionException;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.storage.StorageException;
 import org.apache.ignite.internal.table.distributed.TableMessagesFactory;
 import org.apache.ignite.internal.table.distributed.raft.snapshot.PartitionSnapshotStorage;
 import org.apache.ignite.internal.table.distributed.raft.snapshot.SnapshotUri;
 import org.apache.ignite.internal.table.distributed.raft.snapshot.message.SnapshotMetaRequest;
 import org.apache.ignite.internal.table.distributed.raft.snapshot.message.SnapshotMetaResponse;
+import org.apache.ignite.internal.table.distributed.raft.snapshot.message.SnapshotMvDataResponse;
+import org.apache.ignite.internal.table.distributed.raft.snapshot.message.SnapshotMvDataResponse.ResponseEntry;
+import org.apache.ignite.internal.tostring.IgniteToStringInclude;
+import org.apache.ignite.internal.tostring.S;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.lang.IgniteException;
 import org.apache.ignite.network.ClusterNode;
-import org.apache.ignite.network.MessagingService;
-import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.raft.jraft.Status;
 import org.apache.ignite.raft.jraft.entity.RaftOutter.SnapshotMeta;
-import org.apache.ignite.raft.jraft.storage.SnapshotStorage;
+import org.apache.ignite.raft.jraft.error.RaftError;
 import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotCopier;
 import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotReader;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Snapshot copier implementation for partitions. Used to stream partition data from the leader to the local node.
  */
 public class IncomingSnapshotCopier extends SnapshotCopier {
-    /** Messages factory. */
+    private static final IgniteLogger LOG = Loggers.forClass(IncomingSnapshotCopier.class);
+
     private static final TableMessagesFactory MSG_FACTORY = new TableMessagesFactory();
 
-    /** {@link SnapshotStorage} instance for the partition. */
-    private final PartitionSnapshotStorage snapshotStorage;
+    private static final long NETWORK_TIMEOUT = 10_000;
+
+    private final PartitionSnapshotStorage partitionSnapshotStorage;
 
-    /** Snapshot URI. */
     private final SnapshotUri snapshotUri;
 
-    /** Rebalance thread-pool, used to write data into a storage. */
-    //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-    // Use external pool.
-    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+    private final IgniteSpinBusyLock cancelBusyLock = new IgniteSpinBusyLock();
```

Comment:
Can it be named `busyLock` like everywhere else?
We may have a separate explicit cancellation flag

Reply:
Tried to fix it.